File: Leetcode_go/131_Palindrome_Partitioning.go
package main

func isValid(str string) bool {
	l := 0
	r := len(str) - 1
	for l <= r {
		if str[l] != str[r] {
			return false
		}
		l++
		r--
	}
	return true
}
func backtrack(ans [][]string, temp []string, s string, start int) [][]string {
	if start == len(s) {
		p := make([]string, len(temp))
		copy(p, temp)
		ans := append(ans, p)
		return ans
	}
	for i := start; i < len(s); i++ {
		str := s[start : i+1]
		if !isValid(str) {
			continue
		}
		temp = append(temp, str)
		ans = backtrack(ans, temp, s, i+1)
		temp = temp[:len(temp)-1]
	}
	return ans
}
func partition(s string) [][]string {
	temp := []string{}
	return backtrack([][]string{}, temp, s, 0)
}

File: Leetcode/Backtracking/131_Palindrome_Partitioning.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    bool is_valid(string &s, int l, int r) {
        while (l <= r) {
            if (s[l] != s[r])
                return false;
            l++;
            r--;
        }
        return true;
    }
    void BackPart(vector<vector<string>> &ans, vector<string> &temp, string &s,
                  int start) {
        if (start == s.size()) {
            ans.push_back(temp);
        }
        for (int i = start; i < s.size(); i++) {
            if (!is_valid(s, start, i))
                continue;
            temp.push_back(s.substr(start, i - start + 1));
            BackPart(ans, temp, s, i + 1);
            temp.pop_back();
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> temp;
        BackPart(ans, temp, s, 0);
        return ans;
    }
};

int main() {
    string s = "aab";
    Solution sol;
    auto ans = sol.partition(s);
    for (auto &it : ans) {
        for (auto &i : it)
            cout << i << " ";
        cout << endl;
    }
}
